In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt 
import matplotlib.image as mpimg 
%matplotlib inline
import seaborn as sns

import statsmodels.formula.api as smf
from statsmodels.graphics.gofplots import qqplot 
from scipy import stats

from IPython.display import display
pd.options.display.max_columns = None
#pd.options.display.max_columns = 100
#pd.options.display.max_rows = None --> shows all rows, your notebooks becomes slow 
pd.options.display.max_rows = 200

import warnings
warnings.filterwarnings('ignore')

import os
print (os.getcwd())
#print (os.listdir())

/Users/rajkiran/Desktop/MACHINE LEARNING Praveen/Notebooks


In [2]:
house_df = pd.read_csv("/Users/rajkiran/Desktop/MACHINE LEARNING Praveen/Notebooks/housing_price.csv")
house_df.shape

(1460, 81)

In [3]:
house_df.isnull()

Id  MSSubClass  MSZoning  LotFrontage  LotArea  Street  Alley  \
0     False       False     False        False    False   False   True   
1     False       False     False        False    False   False   True   
2     False       False     False        False    False   False   True   
3     False       False     False        False    False   False   True   
4     False       False     False        False    False   False   True   
...     ...         ...       ...          ...      ...     ...    ...   
1455  False       False     False        False    False   False   True   
1456  False       False     False        False    False   False   True   
1457  False       False     False        False    False   False   True   
1458  False       False     False        False    False   False   True   
1459  False       False     False        False    False   False   True   

      LotShape  LandContour  Utilities  LotConfig  LandSlope  Neighborhood  \
0        False        False      False      False      False         False   
1        False        False      False      False      False         False   
2        False        False      False      False      False         False   
3        False        False      False      False      False         False   
4        False        False      False      False      False         False   
...        ...          ...        ...        ...        ...           ...   
1455     False        False      False      False      False         False   
1456     False        False      False      False      False         False   
1457     False        False      False      False      False         False   
1458     False        False      False      False      False         False   
1459     False        False      False      False      False         False   

      Condition1  Condition2  BldgType  HouseStyle  OverallQual  OverallCond  \
0          False       False     False       False        False        False   
1          False       False     False       False        False        False   
2          False       False     False       False        False        False   
3          False       False     False       False        False        False   
4          False       False     False       False        False        False   
...          ...         ...       ...         ...          ...          ...   
1455       False       False     False       False        False        False   
1456       False       False     False       False        False        False   
1457       False       False     False       False        False        False   
1458       False       False     False       False        False        False   
1459       False       False     False       False        False        False   

      YearBuilt  YearRemodAdd  RoofStyle  RoofMatl  Exterior1st  Exterior2nd  \
0         False         False      False     False        False        False   
1         False         False      False     False        False        False   
2         False         False      False     False        False        False   
3         False         False      False     False        False        False   
4         False         False      False     False        False        False   
...         ...           ...        ...       ...          ...          ...   
1455      False         False      False     False        False        False   
1456      False         False      False     False        False        False   
1457      False         False      False     False        False        False   
1458      False         False      False     False        False        False   
1459      False         False      False     False        False        False   

      MasVnrType  MasVnrArea  ExterQual  ExterCond  Foundation  BsmtQual  \
0          False       False      False      False       False     False   
1          False       False      False      False       False     False   
2          False       False     

In [4]:
house_df = house_df.drop(columns=['Id'])

In [5]:
house_df_infloat = house_df.select_dtypes(['int64','float64'])
house_df_infloat.shape

(1460, 37)

In [6]:
X = house_df_infloat.drop(columns=['SalePrice'])
print(house_df_infloat.shape,X.shape)
y = house_df_infloat['SalePrice']
print(y.shape)

(1460, 37) (1460, 36)
(1460,)


In [7]:
from sklearn.model_selection import train_test_split

In [8]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.3,random_state= 6)

In [9]:
print(X_train.shape,X_test.shape,y_train.shape,y_test.shape)

(1022, 36) (438, 36) (1022,) (438,)


In [10]:
X_train.describe()

MSSubClass  LotFrontage        LotArea  OverallQual  OverallCond  \
count  1022.000000   844.000000    1022.000000  1022.000000  1022.000000   
mean     57.030333    69.386256   10411.110568     6.111546     5.580235   
std      42.010120    23.088927    8758.705203     1.366051     1.070905   
min      20.000000    21.000000    1300.000000     1.000000     2.000000   
25%      20.000000    58.000000    7564.250000     5.000000     5.000000   
50%      50.000000    69.000000    9451.000000     6.000000     5.000000   
75%      70.000000    80.000000   11519.500000     7.000000     6.000000   
max     190.000000   313.000000  164660.000000    10.000000     9.000000   

         YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1   BsmtFinSF2  \
count  1022.000000   1022.000000  1016.000000  1022.000000  1022.000000   
mean   1971.558708   1984.727984    99.052165   448.806262    46.991194   
std      29.880565     20.749954   176.717403   469.060097   166.435944   
min    1880.000000   1950.000000     0.000000     0.000000     0.000000   
25%    1954.000000   1966.000000     0.000000     0.000000     0.000000   
50%    1973.500000   1993.500000     0.000000   383.500000     0.000000   
75%    2000.000000   2004.000000   158.500000   737.250000     0.000000   
max    2010.000000   2010.000000  1600.000000  5644.000000  1474.000000   

         BsmtUnfSF  TotalBsmtSF     1stFlrSF     2ndFlrSF  LowQualFinSF  \
count  1022.000000  1022.000000  1022.000000  1022.000000   1022.000000   
mean    560.158513  1055.955969  1159.887476   344.591977      5.750489   
std     440.874372   447.941476   388.297747   435.772448     48.198815   
min       0.000000     0.000000   334.000000     0.000000      0.000000   
25%     210.500000   796.000000   882.000000     0.000000      0.000000   
50%     464.500000   995.000000  1087.000000     0.000000      0.000000   
75%     809.500000  1305.500000  1389.750000   728.000000      0.000000   
max    2336.000000  6110.000000  4692.000000  2065.000000    572.000000   

         GrLivArea  BsmtFullBath  BsmtHalfBath     FullBath     HalfBath  \
count  1022.000000   1022.000000   1022.000000  1022.000000  1022.000000   
mean   1510.229941      0.429550      0.052838     1.559687     0.377691   
std     521.066682      0.527804      0.223819     0.541933     0.498982   
min     334.000000      0.000000      0.000000     0.000000     0.000000   
25%    1130.250000      0.000000      0.000000     1.000000     0.000000   
50%    1457.000000      0.000000      0.000000     2.000000     0.000000   
75%    1766.750000      1.000000      0.000000     2.000000     1.000000   
max    5642.000000      3.000000      1.000000     3.000000     2.000000   

       BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd   Fireplaces  GarageYrBlt  \
count   1022.000000   1022.000000   1022.000000  1022.000000   971.000000   
mean       2.860078      1.044031      6.508806     0.626223  1978.128733   
std        0.800949      0.209982      1.590071     0.655884    24.886276   
min        0.000000      1.000000      2.000000     0.000000  1906.000000   
25%        2.000000      1.000000      5.000000     0.000000  1961.000000   
50%        3.000000      1.000000      6.000000     1.000000  1980.000000   
75%        3.000000      1.000000      7.000000     1.000000  2001.000000   
max        8.000000      3.000000     14.000000     3.000000  2010.000000   

        GarageCars   GarageArea   WoodDeckSF  OpenPorchSF  EnclosedPorch  \
count  1022.000000  1022.000000  1022.000000  1022.000000    1022.000000   
mean      1.759295   471.094912    96.226027    46.093933      22.340509   
std       0.739793   210.498075   127.090314    64.940936      61.455744   
min       0.000000     0.000000     0.000000     0.000000       0.000000   
25%       1.000000   312.750000     0.000000     0.000000       0.000000   
50%       2.000000   478.500000     0.000000    25.000000       0.000000   
75%       2.000000   576.000000   170.750000    68.000000   

In [11]:
y_train.describe()

count      1022.000000
mean     180922.212329
std       77911.212359
min       34900.000000
25%      130000.000000
50%      163000.000000
75%      214375.000000
max      745000.000000
Name: SalePrice, dtype: float64

In [12]:
X_test.describe()

MSSubClass  LotFrontage        LotArea  OverallQual  OverallCond  \
count  438.000000   357.000000     438.000000   438.000000   438.000000   
mean    56.586758    71.619048   10763.502283     6.070776     5.563927   
std     43.017537    26.870799   12381.226949     1.422929     1.206115   
min     20.000000    21.000000    1477.000000     1.000000     1.000000   
25%     20.000000    60.000000    7500.000000     5.000000     5.000000   
50%     50.000000    70.000000    9593.500000     6.000000     5.000000   
75%     60.000000    81.000000   11622.750000     7.000000     6.000000   
max    190.000000   313.000000  215245.000000    10.000000     9.000000   

         YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1   BsmtFinSF2  \
count   438.000000    438.000000   436.000000   438.000000   438.000000   
mean   1970.589041   1985.187215   114.481651   431.584475    45.518265   
std      30.965990     20.419325   190.588242   424.588869   148.875060   
min    1872.000000   1950.000000     0.000000     0.000000     0.000000   
25%    1951.250000   1968.000000     0.000000     0.000000     0.000000   
50%    1971.500000   1994.000000     0.000000   382.500000     0.000000   
75%    2002.000000   2004.000000   178.500000   683.750000     0.000000   
max    2009.000000   2010.000000  1170.000000  2260.000000  1085.000000   

         BsmtUnfSF  TotalBsmtSF     1stFlrSF     2ndFlrSF  LowQualFinSF  \
count   438.000000   438.000000   438.000000   438.000000    438.000000   
mean    583.764840  1060.867580  1169.018265   352.593607      6.063927   
std     444.240191   416.836567   382.933056   438.735612     49.654002   
min       0.000000     0.000000   372.000000     0.000000      0.000000   
25%     245.250000   795.250000   886.000000     0.000000      0.000000   
50%     505.000000   977.000000  1087.000000     0.000000      0.000000   
75%     797.750000  1281.500000  1404.250000   728.000000      0.000000   
max    2153.000000  3138.000000  3138.000000  1872.000000    513.000000   

         GrLivArea  BsmtFullBath  BsmtHalfBath    FullBath    HalfBath  \
count   438.000000    438.000000    438.000000  438.000000  438.000000   
mean   1527.675799      0.415525      0.068493    1.577626    0.394977   
std     536.041111      0.497992      0.270372    0.571763    0.512250   
min     480.000000      0.000000      0.000000    0.000000    0.000000   
25%    1130.000000      0.000000      0.000000    1.000000    0.000000   
50%    1473.000000      0.000000      0.000000    2.000000    0.000000   
75%    1800.000000      1.000000      0.000000    2.000000    1.000000   
max    4676.000000      2.000000      2.000000    3.000000    2.000000   

       BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  GarageYrBlt  \
count    438.000000    438.000000    438.000000  438.000000   408.000000   
mean       2.881279      1.052511      6.538813    0.582192  1979.404412   
std        0.850127      0.242943      1.706657    0.617334    24.222062   
min        0.000000      0.000000      3.000000    0.000000  1900.000000   
25%        2.000000      1.000000      5.000000    0.000000  1963.000000   
50%        3.000000      1.000000      6.000000    1.000000  1980.000000   
75%        3.000000      1.000000      7.000000    1.000000  2003.000000   
max        6.000000      3.000000     12.000000    2.000000  2009.000000   

       GarageCars   GarageArea  WoodDeckSF  OpenPorchSF  EnclosedPorch  \
count  438.000000   438.000000  438.000000   438.000000     438.000000   
mean     1.785388   477.378995   89.621005    47.981735      21.052511   
std      0.765124   221.511900  121.169955    69.285762      60.386584   
min      0.000000     0.000000    0.000000     0.000000       0.000000   
25%      1.000000   352.000000    0.000000     0.000000       0.000000   
50%      2.000000   480.000000    0.000000    24.000000       0.000000   
75%      2.000000   588.000000  165.750000    66.000000       0.000000   
max      4.000000  1390.000000  574

In [13]:
y_test.describe()

count       438.000000
mean     180918.824201
std       82997.077815
min       35311.000000
25%      129600.000000
50%      162250.000000
75%      210750.000000
max      755000.000000
Name: SalePrice, dtype: float64

In [14]:
def missing_zero_values_table(df):
        zero_val = (df == 0.00).astype(int).sum(axis=0)
        mis_val = df.isnull().sum()
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        mz_table = pd.concat([zero_val, mis_val, mis_val_percent], axis=1)
        mz_table = mz_table.rename(
        columns = {0 : 'Zero Values', 1 : 'Missing Values', 2 : '% of Total Values'})
        mz_table['Total Zero Missing Values'] = mz_table['Zero Values'] + mz_table['Missing Values']
        mz_table['% Total Zero Missing Values'] = 100 * mz_table['Total Zero Missing Values'] / len(df)
        mz_table['Data Type'] = df.dtypes
        mz_table = mz_table[
            mz_table.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns and " + str(df.shape[0]) + " Rows.\n"      
            "There are " + str(mz_table.shape[0]) +
              " columns that have missing values.")
#         mz_table.to_excel('D:/sampledata/missing_and_zero_values.xlsx', freeze_panes=(1,0), index = False)
        return mz_table

#https://stackoverflow.com/questions/51070985/find-out-the-percentage-of-missing-values-in-each-column-in-the-given-dataset

In [15]:
missing_zero_values_table(X_train)

Your selected dataframe has 36 columns and 1022 Rows.
There are 3 columns that have missing values.


Zero Values  Missing Values  % of Total Values  \
LotFrontage            0             178               17.4   
GarageYrBlt            0              51                5.0   
MasVnrArea           617               6                0.6   

             Total Zero Missing Values  % Total Zero Missing Values Data Type  
LotFrontage                        178                         17.4   float64  
GarageYrBlt                         51                          5.0   float64  
MasVnrArea                         623                         61.0   float64

In [16]:
X_train.shape

(1022, 36)

In [17]:
X_train[['MasVnrArea']]

MasVnrArea
435          0.0
1166        72.0
935        108.0
588          0.0
80         576.0
...          ...
1389         0.0
618        452.0
227        127.0
713        160.0
906        208.0

[1022 rows x 1 columns]

In [18]:
X_train['MasVnrArea'].fillna(X_train['MasVnrArea'].mean(),inplace=True)

In [19]:
X_train[['LotFrontage']].describe()   # 69.4

LotFrontage
count   844.000000
mean     69.386256
std      23.088927
min      21.000000
25%      58.000000
50%      69.000000
75%      80.000000
max     313.000000

In [20]:
X_train['LotFrontage'].fillna(X_train['LotFrontage'].mean(),inplace=True)

In [21]:
X_train['LotFrontage'].describe()

count    1022.000000
mean       69.386256
std        20.979961
min        21.000000
25%        60.000000
50%        69.386256
75%        78.000000
max       313.000000
Name: LotFrontage, dtype: float64

In [22]:
X_train[['GarageYrBlt']].describe()   # 69.4

GarageYrBlt
count   971.000000
mean   1978.128733
std      24.886276
min    1906.000000
25%    1961.000000
50%    1980.000000
75%    2001.000000
max    2010.000000

In [23]:
X_train['GarageYrBlt'].fillna(X_train['GarageYrBlt'].mean(),inplace=True)

In [24]:
X_train[['GarageYrBlt']].describe()   # 69.4

GarageYrBlt
count  1022.000000
mean   1978.128733
std      24.256766
min    1906.000000
25%    1962.000000
50%    1978.128733
75%    2000.750000
max    2010.000000

In [25]:
def missing_zero_values_table(df):
        zero_val = (df == 0.00).astype(int).sum(axis=0)
        mis_val = df.isnull().sum()
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        mz_table = pd.concat([zero_val, mis_val, mis_val_percent], axis=1)
        mz_table = mz_table.rename(
        columns = {0 : 'Zero Values', 1 : 'Missing Values', 2 : '% of Total Values'})
        mz_table['Total Zero Missing Values'] = mz_table['Zero Values'] + mz_table['Missing Values']
        mz_table['% Total Zero Missing Values'] = 100 * mz_table['Total Zero Missing Values'] / len(df)
        mz_table['Data Type'] = df.dtypes
        mz_table = mz_table[
            mz_table.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns and " + str(df.shape[0]) + " Rows.\n"      
            "There are " + str(mz_table.shape[0]) +
              " columns that have missing values.")
#         mz_table.to_excel('D:/sampledata/missing_and_zero_values.xlsx', freeze_panes=(1,0), index = False)
        return mz_table

#https://stackoverflow.com/questions/51070985/find-out-the-percentage-of-missing-values-in-each-column-in-the-given-dataset

In [26]:
missing_zero_values_table(X_train)

Your selected dataframe has 36 columns and 1022 Rows.
There are 0 columns that have missing values.


Empty DataFrame
Columns: [Zero Values, Missing Values, % of Total Values, Total Zero Missing Values, % Total Zero Missing Values, Data Type]
Index: []

In [27]:
missing_zero_values_table(X_test)

Your selected dataframe has 36 columns and 438 Rows.
There are 3 columns that have missing values.


Zero Values  Missing Values  % of Total Values  \
LotFrontage            0              81               18.5   
GarageYrBlt            0              30                6.8   
MasVnrArea           244               2                0.5   

             Total Zero Missing Values  % Total Zero Missing Values Data Type  
LotFrontage                         81                         18.5   float64  
GarageYrBlt                         30                          6.8   float64  
MasVnrArea                         246                         56.2   float64

In [28]:
X_test['LotFrontage'].fillna(X_test['LotFrontage'].mean(),inplace=True)

In [29]:
X_test['GarageYrBlt'].fillna(X_test['GarageYrBlt'].mean(),inplace=True)

In [30]:
X_test['MasVnrArea'].fillna(X_test['MasVnrArea'].mean(),inplace=True)

In [31]:
missing_zero_values_table(X_test)

Your selected dataframe has 36 columns and 438 Rows.
There are 0 columns that have missing values.


Empty DataFrame
Columns: [Zero Values, Missing Values, % of Total Values, Total Zero Missing Values, % Total Zero Missing Values, Data Type]
Index: []

In [32]:
y_test = pd.DataFrame(y_test)
missing_zero_values_table(y_test)

Your selected dataframe has 1 columns and 438 Rows.
There are 0 columns that have missing values.


Empty DataFrame
Columns: [Zero Values, Missing Values, % of Total Values, Total Zero Missing Values, % Total Zero Missing Values, Data Type]
Index: []

# Linear Reg

In [33]:
from sklearn.linear_model import LinearRegression
%time
lm_model = LinearRegression() #create an object for "LinearRegression()" ML algorithm 
lm_model.fit(X_train, y_train) #train the algorithm on data

CPU times: user 3 µs, sys: 1e+03 ns, total: 4 µs
Wall time: 9.3 µs


LinearRegression()

In [34]:
lm_model.score(X_train,y_train)

0.8167150742484676

In [35]:
from sklearn.metrics import mean_squared_error, r2_score
#y_hat, predictions of "saleprice" over train set.
pred_train = lm_model.predict(X_train) #make the predictions over train set
mse_train = mean_squared_error(y_train, pred_train)# y (actual sale price) and y_hat (predicted saleprice) 
r2_train = r2_score(y_train, pred_train)
print('Root Mean Squared Error on Train Set : ', np.sqrt(mse_train))
print('R-Squre on Train Set : ', r2_train)

Root Mean Squared Error on Train Set :  33338.86108494312
R-Squre on Train Set :  0.8167150742484676


# DT

In [36]:
from sklearn.tree import DecisionTreeRegressor
%time
lmDTR_model = DecisionTreeRegressor(random_state=6) #create an object for "LinearRegression()" ML algorithm 
lmDTR_model.fit(X_train, y_train) #train the algorithm on data

CPU times: user 4 µs, sys: 0 ns, total: 4 µs
Wall time: 7.87 µs


DecisionTreeRegressor(random_state=6)

In [37]:
lmDTR_model.score(X_train,y_train)

1.0

In [38]:
from sklearn.metrics import mean_squared_error, r2_score
#y_hat, predictions of "saleprice" over train set.
pred_train = lmDTR_model.predict(X_train) #make the predictions over train set
mse_train = mean_squared_error(y_train, pred_train)# y (actual sale price) and y_hat (predicted saleprice) 
r2_train = r2_score(y_train, pred_train)
print('Root Mean Squared Error on Train Set : ', np.sqrt(mse_train))
print('R-Squre on Train Set : ', r2_train)

Root Mean Squared Error on Train Set :  0.0
R-Squre on Train Set :  1.0


# Stochastic GD

In [39]:
from sklearn.linear_model import SGDRegressor
%time
lmSGD_model = SGDRegressor(random_state=6) #create an object for "LinearRegression()" ML algorithm 
lmSGD_model.fit(X_train, y_train) #train the algorithm on data

CPU times: user 3 µs, sys: 1e+03 ns, total: 4 µs
Wall time: 9.06 µs


SGDRegressor(random_state=6)

In [40]:
lmSGD_model.score(X_train,y_train)

-5.1648612488999885e+23

In [41]:
from sklearn.metrics import mean_squared_error, r2_score
#y_hat, predictions of "saleprice" over train set.
pred_train = lmSGD_model.predict(X_train) #make the predictions over train set
mse_train = mean_squared_error(y_train, pred_train)# y (actual sale price) and y_hat (predicted saleprice) 
r2_train = r2_score(y_train, pred_train)
print('Root Mean Squared Error on Train Set : ', np.sqrt(mse_train))
print('R-Squre on Train Set : ', r2_train)

Root Mean Squared Error on Train Set :  5.596502665893825e+16
R-Squre on Train Set :  -5.1648612488999885e+23


# Linear SVR

In [42]:
from sklearn.svm import LinearSVR
%time
lmSVR_model = LinearSVR(random_state=6) #create an object for "LinearRegression()" ML algorithm 
lmSVR_model.fit(X_train, y_train) #train the algorithm on data

CPU times: user 5 µs, sys: 2 µs, total: 7 µs
Wall time: 16.2 µs


LinearSVR(random_state=6)

In [43]:
lmSVR_model.score(X_train,y_train)

0.47528424167627825

In [44]:
from sklearn.metrics import mean_squared_error, r2_score
#y_hat, predictions of "saleprice" over train set.
pred_train = lmSVR_model.predict(X_train) #make the predictions over train set
mse_train = mean_squared_error(y_train, pred_train)# y (actual sale price) and y_hat (predicted saleprice) 
r2_train = r2_score(y_train, pred_train)
print('Root Mean Squared Error on Train Set : ', np.sqrt(mse_train))
print('R-Squre on Train Set : ', r2_train)

Root Mean Squared Error on Train Set :  56409.13486495055
R-Squre on Train Set :  0.47528424167627825


# Random Forest

In [45]:
from sklearn.ensemble import RandomForestRegressor
%time
lmRFR_model = RandomForestRegressor(random_state=6) #create an object for "LinearRegression()" ML algorithm 
lmRFR_model.fit(X_train, y_train) #train the algorithm on data

CPU times: user 3 µs, sys: 1e+03 ns, total: 4 µs
Wall time: 6.91 µs


RandomForestRegressor(random_state=6)

In [46]:
lmRFR_model.score(X_train,y_train)

0.9791628828607899

In [224]:
from sklearn.metrics import mean_squared_error, r2_score
#y_hat, predictions of "saleprice" over train set.
pred_train = lmRFR_model.predict(X_train) #make the predictions over train set
mse_train = mean_squared_error(y_train, pred_train)# y (actual sale price) and y_hat (predicted saleprice) 
r2_train = r2_score(y_train, pred_train)
print('Root Mean Squared Error on Train Set : ', np.sqrt(mse_train))
print('R-Squre on Train Set : ', r2_train)

Root Mean Squared Error on Train Set :  11241.03246521659
R-Squre on Train Set :  0.9791628828607899


# Gradient Boosting

In [48]:
from sklearn.ensemble import GradientBoostingRegressor
%time
lmGBR_model = GradientBoostingRegressor(random_state=6) #create an object for "LinearRegression()" ML algorithm 
lmGBR_model.fit(X_train, y_train) #train the algorithm on data

CPU times: user 4 µs, sys: 1e+03 ns, total: 5 µs
Wall time: 9.06 µs


GradientBoostingRegressor(random_state=6)

In [49]:
lmGBR_model.score(X_train,y_train)

0.9655169054442602

In [50]:
from sklearn.metrics import mean_squared_error, r2_score
#y_hat, predictions of "saleprice" over train set.
pred_train = lmGBR_model.predict(X_train) #make the predictions over train set
mse_train = mean_squared_error(y_train, pred_train)# y (actual sale price) and y_hat (predicted saleprice) 
r2_train = r2_score(y_train, pred_train)
print('Root Mean Squared Error on Train Set : ', np.sqrt(mse_train))
print('R-Squre on Train Set : ', r2_train)

Root Mean Squared Error on Train Set :  14460.739478035255
R-Squre on Train Set :  0.9655169054442602


# Lasso

In [51]:
from sklearn.linear_model import Lasso
%time
lmLasso_model = Lasso(random_state=6) #create an object for "LinearRegression()" ML algorithm 
lmLasso_model.fit(X_train, y_train) #train the algorithm on data

CPU times: user 4 µs, sys: 1 µs, total: 5 µs
Wall time: 8.82 µs


Lasso(random_state=6)

In [52]:
lmLasso_model.score(X_train,y_train)

0.8167150556066979

In [53]:
from sklearn.metrics import mean_squared_error, r2_score
#y_hat, predictions of "saleprice" over train set.
pred_train = lmLasso_model.predict(X_train) #make the predictions over train set
mse_train = mean_squared_error(y_train, pred_train)# y (actual sale price) and y_hat (predicted saleprice) 
r2_train = r2_score(y_train, pred_train)
print('Root Mean Squared Error on Train Set : ', np.sqrt(mse_train))
print('R-Squre on Train Set : ', r2_train)

Root Mean Squared Error on Train Set :  33338.86278037811
R-Squre on Train Set :  0.8167150556066979


# KNN

In [54]:
from sklearn.neighbors import KNeighborsRegressor
%time
lmknn_model = KNeighborsRegressor() #create an object for "LinearRegression()" ML algorithm 
lmknn_model.fit(X_train, y_train) #train the algorithm on data

CPU times: user 3 µs, sys: 1e+03 ns, total: 4 µs
Wall time: 9.06 µs


KNeighborsRegressor()

In [55]:
lmknn_model.score(X_train,y_train)

0.7709443105050775

In [56]:
from sklearn.metrics import mean_squared_error, r2_score
#y_hat, predictions of "saleprice" over train set.
pred_train = lmknn_model.predict(X_train) #make the predictions over train set
mse_train = mean_squared_error(y_train, pred_train)# y (actual sale price) and y_hat (predicted saleprice) 
r2_train = r2_score(y_train, pred_train)
print('Root Mean Squared Error on Train Set : ', np.sqrt(mse_train))
print('R-Squre on Train Set : ', r2_train)

Root Mean Squared Error on Train Set :  37269.87424145846
R-Squre on Train Set :  0.7709443105050775


# Ridge

from sklearn.linear_model import ridge_regression
%time
lmRReg_model = ridge_regression(X=X_train,y=y_train,alpha=0.01) #create an object for "LinearRegression()" ML algorithm 
#lmRReg_model.fit(X_train, y_train) #train the algorithm on data

lmRReg_model.score(X_train,y_train)

from sklearn.metrics import mean_squared_error, r2_score
#y_hat, predictions of "saleprice" over train set.
pred_train = lmRReg_model.predict(X_train) #make the predictions over train set
mse_train = mean_squared_error(y_train, pred_train)# y (actual sale price) and y_hat (predicted saleprice) 
r2_train = r2_score(y_train, pred_train)
print('Root Mean Squared Error on Train Set : ', np.sqrt(mse_train))
print('R-Squre on Train Set : ', r2_train)

In [57]:
def missing_zero_values_table(df):
        zero_val = (df == 0.00).astype(int).sum(axis=0)
        mis_val = df.isnull().sum()
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        mz_table = pd.concat([zero_val, mis_val, mis_val_percent], axis=1)
        mz_table = mz_table.rename(
        columns = {0 : 'Zero Values', 1 : 'Missing Values', 2 : '% of Total Values'})
        mz_table['Total Zero Missing Values'] = mz_table['Zero Values'] + mz_table['Missing Values']
        mz_table['% Total Zero Missing Values'] = 100 * mz_table['Total Zero Missing Values'] / len(df)
        mz_table['Data Type'] = df.dtypes
        mz_table = mz_table[
            mz_table.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns and " + str(df.shape[0]) + " Rows.\n"      
            "There are " + str(mz_table.shape[0]) +
              " columns that have missing values.")
#         mz_table.to_excel('D:/sampledata/missing_and_zero_values.xlsx', freeze_panes=(1,0), index = False)
        return mz_table

#https://stackoverflow.com/questions/51070985/find-out-the-percentage-of-missing-values-in-each-column-in-the-given-dataset

In [58]:
y_test = pd.DataFrame(y_test)
type(y_test)

pandas.core.frame.DataFrame

In [59]:
missing_zero_values_table(y_test)

Your selected dataframe has 1 columns and 438 Rows.
There are 0 columns that have missing values.


Empty DataFrame
Columns: [Zero Values, Missing Values, % of Total Values, Total Zero Missing Values, % Total Zero Missing Values, Data Type]
Index: []

In [60]:
def evaluate_regressor(xt,yt,model):
    pred_yt = model.predict(xt)
    mse_train = mean_squared_error(yt, pred_yt)# y (actual sale price) and y_hat (predicted saleprice) r2_train = r2_score(yt, pred_yt)
    return np.sqrt(mse_train), r2_train

## RMSE , R**2 for LR

In [61]:
rmse,r2 = evaluate_regressor(X_train,y_train,lm_model) 
print (rmse,r2)

33338.86108494312 0.7709443105050775


## Rmse and R**2 for DT

In [62]:
rmse,r2 = evaluate_regressor(X_train,y_train,lmDTR_model) 
print (rmse,r2)

0.0 0.7709443105050775


## Rmse and R**2 for SGD


In [63]:
rmse,r2 = evaluate_regressor(X_train,y_train,lmSGD_model) 
print (rmse,r2)

5.596502665893825e+16 0.7709443105050775


## Rmse and R**2 for Linear SVR


In [64]:
rmse,r2 = evaluate_regressor(X_train,y_train,lmSVR_model) 
print (rmse,r2)

56409.13486495055 0.7709443105050775


## RMSE and R**2 for Random Forest

In [65]:
rmse,r2 = evaluate_regressor(X_train,y_train,lmRFR_model) 
print (rmse,r2)

11241.03246521659 0.7709443105050775


## RMSE and R**2 for Gradient Boosting

In [66]:
rmse,r2 = evaluate_regressor(X_train,y_train,lmGBR_model) 
print (rmse,r2)

14460.739478035255 0.7709443105050775


## RMSE and R**2 for Lasso

In [67]:
rmse,r2 = evaluate_regressor(X_train,y_train,lmLasso_model) 
print (rmse,r2)

33338.86278037811 0.7709443105050775


## RMSE and R**2 for Knn

In [68]:
rmse,r2 = evaluate_regressor(X_train,y_train,lmknn_model) 
print (rmse,r2)

37269.87424145846 0.7709443105050775


# K-folds Cross val

In [69]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold


# For LR

In [70]:
# create a KFold object with 10 splits
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
scores = cross_val_score(lm_model, X_train, y_train, scoring='r2', cv=folds)
print(scores)
print("\n")
print(scores.mean())

[0.83008304 0.74459879 0.13207514 0.86197699 0.87145095 0.73032646
 0.88340981 0.84213544 0.82550063 0.84744392]


0.7569001170702236


In [71]:
# can tune other metrics, such as MSE
scores = cross_val_score(lm_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds) 
scores

array([-1.25952567e+09, -2.19499226e+09, -6.13403534e+09, -6.27208283e+08,
       -7.03305928e+08, -1.64786629e+09, -5.51222188e+08, -5.02489579e+08,
       -1.25823439e+09, -8.88182952e+08])

In [72]:
#RMSEs
np.sqrt(abs(scores))

array([35489.79670882, 46850.74450283, 78320.08260822, 25044.12670975,
       26519.91568095, 40593.91932049, 23478.12146724, 22416.27932049,
       35471.59983099, 29802.3984276 ])

In [73]:
np.sqrt(abs(scores)).mean()

36398.69845773642

In [74]:
rmse,r2 = evaluate_regressor(X_test,y_test,lm_model)
print (rmse,r2)

37568.90804668851 0.7709443105050775


In [75]:
print ("Training Scores: ")
rmse,r2 = evaluate_regressor(X_train,y_train,lm_model) 
print (rmse,r2)

print ("\nCV Scores: ")
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
r2_scores = cross_val_score(lm_model, X_train, y_train, scoring='r2', cv=folds)
rmse_scores = cross_val_score(lm_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds)
print (np.sqrt(abs(scores)).mean(),r2_scores.mean())

print("\n Test Scores: ")
rmse,r2 = evaluate_regressor(X_test,y_test,lm_model) 
print (rmse,r2)

Training Scores: 
33338.86108494312 0.7709443105050775

CV Scores: 
36398.69845773642 0.7569001170702236

 Test Scores: 
37568.90804668851 0.7709443105050775


## For DT

In [76]:
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
scores = cross_val_score(lmDTR_model, X_train, y_train, scoring='r2', cv=folds)
print(scores)
print("\n")
print(scores.mean())

[0.70508817 0.85646142 0.76995403 0.60622016 0.6081359  0.63276135
 0.80216803 0.77484678 0.7854765  0.74266876]


0.7283781096775638


In [77]:
# can tune other metrics, such as MSE
scores = cross_val_score(lmDTR_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds) 
scores

array([-2.18606203e+09, -1.23361227e+09, -1.62584361e+09, -1.78942617e+09,
       -2.14393143e+09, -2.24404736e+09, -9.35321991e+08, -7.16672259e+08,
       -1.54682991e+09, -1.49818492e+09])

In [78]:
#RMSEs
np.sqrt(abs(scores))

array([46755.34229462, 35122.81694795, 40321.75111804, 42301.60954083,
       46302.60719282, 47371.37702759, 30583.03436756, 26770.73511556,
       39329.75856442, 38706.39377859])

In [79]:
#RMSEs
np.sqrt(abs(scores)).mean()

39356.542594799146

In [80]:
rmse,r2 = evaluate_regressor(X_test,y_test,lmDTR_model)
print (rmse,r2)

49397.45892811486 0.7709443105050775


In [81]:
print ("Training Scores: ")
rmse,r2 = evaluate_regressor(X_train,y_train,lmDTR_model) 
print (rmse,r2)

print ("\nCV Scores: ")
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
r2_scores = cross_val_score(lmDTR_model, X_train, y_train, scoring='r2', cv=folds)
rmse_scores = cross_val_score(lmDTR_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds)
print (np.sqrt(abs(scores)).mean(),r2_scores.mean())

print("\n Test Scores: ")
rmse,r2 = evaluate_regressor(X_test,y_test,lmDTR_model) 
print (rmse,r2)

Training Scores: 
0.0 0.7709443105050775

CV Scores: 
39356.542594799146 0.7283781096775638

 Test Scores: 
49397.45892811486 0.7709443105050775


# For SGD

In [82]:
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
scores = cross_val_score(lmSGD_model, X_train, y_train, scoring='r2', cv=folds)
print(scores)
print("\n")
print(scores.mean())

[-6.06184436e+24 -7.90020704e+23 -1.14629127e+24 -1.04158244e+24
 -1.20983039e+22 -2.06466186e+25 -5.69474858e+23 -3.78821846e+23
 -4.90559081e+23 -1.24570026e+23]


-3.126188151342469e+24


In [83]:
# can tune other metrics, such as MSE
scores = cross_val_score(lmSGD_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds)
scores

array([-4.49339987e+34, -6.78966754e+33, -8.10138236e+33, -4.73319015e+33,
       -6.61911458e+31, -1.26163164e+35, -2.69239785e+33, -1.20580600e+33,
       -3.53719506e+33, -7.25247877e+32])

In [84]:
#RMSEs
np.sqrt(abs(scores))

array([2.11976411e+17, 8.23994389e+16, 9.00076794e+16, 6.87981842e+16,
       8.13579411e+15, 3.55194543e+17, 5.18883209e+16, 3.47247175e+16,
       5.94743227e+16, 2.69304266e+16])

In [85]:
#RMSEs
np.sqrt(abs(scores)).mean()

9.895298383316957e+16

In [86]:
rmse,r2 = evaluate_regressor(X_test,y_test,lmSGD_model)
print (rmse,r2)

7.901322162653406e+16 0.7709443105050775


In [87]:
print ("Training Scores: ")
rmse,r2 = evaluate_regressor(X_train,y_train,lmSGD_model) 
print (rmse,r2)

print ("\nCV Scores: ")
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
r2_scores = cross_val_score(lmSGD_model, X_train, y_train, scoring='r2', cv=folds)
rmse_scores = cross_val_score(lmSGD_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds)
print (np.sqrt(abs(scores)).mean(),r2_scores.mean())

print("\n Test Scores: ")
rmse,r2 = evaluate_regressor(X_test,y_test,lmSGD_model) 
print (rmse,r2)

Training Scores: 
5.596502665893825e+16 0.7709443105050775

CV Scores: 
9.895298383316957e+16 -3.126188151342469e+24

 Test Scores: 
7.901322162653406e+16 0.7709443105050775


# For Linear SVR

In [88]:
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
scores = cross_val_score(lmSVR_model, X_train, y_train, scoring='r2', cv=folds)
print(scores)
print("\n")
print(scores.mean())

[ 0.66336612  0.3718413  -0.17460038  0.52539442  0.75747797  0.51933279
  0.84223347  0.6565425   0.80035646  0.78657933]


0.5748523968684472


In [89]:
# can tune other metrics, such as MSE
scores = cross_val_score(lmSVR_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds)
scores

array([-2.49533071e+09, -5.39857840e+09, -8.30145620e+09, -2.15671691e+09,
       -1.32686461e+09, -2.93716358e+09, -7.45898188e+08, -1.09323981e+09,
       -1.43953738e+09, -1.24253719e+09])

In [90]:
#RMSEs
np.sqrt(abs(scores))

array([49953.28529968, 73475.01890455, 91112.32737936, 46440.46631761,
       36426.15279786, 54195.6048189 , 27311.13670324, 33064.1770976 ,
       37941.23591343, 35249.64097375])

In [91]:
#RMSEs
np.sqrt(abs(scores)).mean()

48516.904620598

In [92]:
rmse,r2 = evaluate_regressor(X_test,y_test,lmSVR_model)
print (rmse,r2)

57363.39135599983 0.7709443105050775


In [93]:
print ("Training Scores: ")
rmse,r2 = evaluate_regressor(X_train,y_train,lmSVR_model) 
print (rmse,r2)

print ("\nCV Scores: ")
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
r2_scores = cross_val_score(lmSVR_model, X_train, y_train, scoring='r2', cv=folds)
rmse_scores = cross_val_score(lmSVR_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds)
print (np.sqrt(abs(scores)).mean(),r2_scores.mean())

print("\n Test Scores: ")
rmse,r2 = evaluate_regressor(X_test,y_test,lmSVR_model) 
print (rmse,r2)

Training Scores: 
56409.13486495055 0.7709443105050775

CV Scores: 
48516.904620598 0.5748523968684472

 Test Scores: 
57363.39135599983 0.7709443105050775


# Random Forest

In [94]:
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
scores = cross_val_score(lmRFR_model, X_train, y_train, scoring='r2', cv=folds)
print(scores)
print("\n")
print(scores.mean())

[0.85760846 0.86358179 0.65680425 0.8328593  0.89668219 0.78668309
 0.91035975 0.91768288 0.89727251 0.83282275]


0.8452356962351916


In [95]:
# can tune other metrics, such as MSE
scores = cross_val_score(lmRFR_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds)
scores

array([-1.05549085e+09, -1.17241773e+09, -2.42552659e+09, -7.59525800e+08,
       -5.65263069e+08, -1.30349365e+09, -4.23806645e+08, -2.62018896e+08,
       -7.40720497e+08, -9.73307535e+08])

In [96]:
#RMSEs
np.sqrt(abs(scores))

array([32488.31859551, 34240.58595819, 49249.63543456, 27559.49564666,
       23775.26169441, 36103.92850773, 20586.56467633, 16186.99774783,
       27216.18078891, 31197.87708514])

In [97]:
#RMSEs
np.sqrt(abs(scores)).mean()

29860.484613525612

In [98]:
rmse,r2 = evaluate_regressor(X_test,y_test,lmRFR_model)
print (rmse,r2)

33405.99252000547 0.7709443105050775


In [99]:
print ("Training Scores: ")
rmse,r2 = evaluate_regressor(X_train,y_train,lmRFR_model) 
print (rmse,r2)

print ("\nCV Scores: ")
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
r2_scores = cross_val_score(lmRFR_model, X_train, y_train, scoring='r2', cv=folds)
rmse_scores = cross_val_score(lmRFR_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds)
print (np.sqrt(abs(scores)).mean(),r2_scores.mean())

print("\n Test Scores: ")
rmse,r2 = evaluate_regressor(X_test,y_test,lmRFR_model) 
print (rmse,r2)

Training Scores: 
11241.03246521659 0.7709443105050775

CV Scores: 
29860.484613525612 0.8452356962351916

 Test Scores: 
33405.99252000547 0.7709443105050775


# Gradient Boosting

In [100]:
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
scores = cross_val_score(lmGBR_model, X_train, y_train, scoring='r2', cv=folds)
print(scores)
print("\n")
print(scores.mean())

[0.89694879 0.89550863 0.59549172 0.88647072 0.90160192 0.8411705
 0.91991998 0.91730883 0.89110602 0.88625838]


0.8631785487564585


In [101]:
# can tune other metrics, such as MSE
scores = cross_val_score(lmGBR_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds) 
scores

array([-7.63876940e+08, -8.98029190e+08, -2.85885124e+09, -5.15903167e+08,
       -5.38346682e+08, -9.70543069e+08, -3.78607189e+08, -2.63209497e+08,
       -7.85184197e+08, -6.62204789e+08])

In [102]:
#RMSEs
np.sqrt(abs(scores))

array([27638.32375187, 29967.13516531, 53468.22642356, 22713.50185549,
       23202.29906963, 31153.54023236, 19457.83104991, 16223.7325297 ,
       28021.13839628, 25733.34003423])

In [103]:
#RMSEs
np.sqrt(abs(scores)).mean()

27757.906850834

In [104]:
rmse,r2 = evaluate_regressor(X_test,y_test,lmGBR_model)
print (rmse,r2)

30472.896344347435 0.7709443105050775


In [105]:
print ("Training Scores: ")
rmse,r2 = evaluate_regressor(X_train,y_train,lmGBR_model) 
print (rmse,r2)

print ("\nCV Scores: ")
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
r2_scores = cross_val_score(lmGBR_model, X_train, y_train, scoring='r2', cv=folds)
rmse_scores = cross_val_score(lmGBR_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds)
print (np.sqrt(abs(scores)).mean(),r2_scores.mean())

print("\n Test Scores: ")
rmse,r2 = evaluate_regressor(X_test,y_test,lmGBR_model) 
print (rmse,r2)

Training Scores: 
14460.739478035255 0.7709443105050775

CV Scores: 
27757.906850834 0.8631785487564585

 Test Scores: 
30472.896344347435 0.7709443105050775


# Lasso

In [106]:
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
scores = cross_val_score(lmLasso_model, X_train, y_train, scoring='r2', cv=folds)
print(scores)
print("\n")
print(scores.mean())

[0.83008251 0.7446067  0.13212219 0.8619583  0.871457   0.73034592
 0.88344149 0.84217018 0.82552156 0.8474583 ]


0.7569164136770342


In [107]:
# can tune other metrics, such as MSE
scores = cross_val_score(lmLasso_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds) 
scores

array([-1.25952959e+09, -2.19492424e+09, -6.13370278e+09, -6.27293247e+08,
       -7.03272850e+08, -1.64774741e+09, -5.51072419e+08, -5.02379001e+08,
       -1.25808350e+09, -8.88099236e+08])

In [108]:
#RMSEs
np.sqrt(abs(scores))

array([35489.85198409, 46850.0186108 , 78317.95952288, 25045.82295445,
       26519.29202751, 40592.45508517, 23474.9317155 , 22413.81272462,
       35469.47275379, 29800.99387719])

In [109]:
#RMSEs
np.sqrt(abs(scores)).mean()

36397.461125601

In [110]:
rmse,r2 = evaluate_regressor(X_test,y_test,lmLasso_model)
print (rmse,r2)

37569.45954171043 0.7709443105050775


In [111]:
print ("Training Scores: ")
rmse,r2 = evaluate_regressor(X_train,y_train,lmLasso_model) 
print (rmse,r2)

print ("\nCV Scores: ")
folds = KFold(n_splits = 10, shuffle = True, random_state = 10)
r2_scores = cross_val_score(lmLasso_model, X_train, y_train, scoring='r2', cv=folds)
rmse_scores = cross_val_score(lmLasso_model, X_train, y_train, scoring='neg_mean_squared_error', cv=folds)
print (np.sqrt(abs(scores)).mean(),r2_scores.mean())

print("\n Test Scores: ")
rmse,r2 = evaluate_regressor(X_test,y_test,lmLasso_model) 
print (rmse,r2)

Training Scores: 
33338.86278037811 0.7709443105050775

CV Scores: 
36397.461125601 0.7569164136770342

 Test Scores: 
37569.45954171043 0.7709443105050775


In [115]:
X_train.tail()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
1389          50         60.0     6000            6            6       1941   
618           20         90.0    11694            9            5       2007   
227          160         21.0     1869            6            6       1970   
713          190         60.0     9873            4            5       1970   
906           20        116.0    13501            8            5       2006   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  \
1389          1950         0.0         375           0        360   
618           2007       452.0          48           0       1774   
227           1970       127.0         321           0        162   
713           1970       160.0         789           0        171   
906           2006       208.0          63           0       1560   

      TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
1389          735       869       349             0       1218             0   
618          1822      1828         0             0       1828             0   
227           483       483       504             0        987             0   
713           960       960         0             0        960             1   
906          1623      1636         0             0       1636             1   

      BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
1389             1         1         0             3             1   
618              0         2         0             3             1   
227              0         1         1             2             1   
713              0         1         0             3             1   
906              0         2         0             3             1   

      TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
1389             6           1       2003.0           2         440   
618              9           1       2007.0           3         774   
227              5           0       1987.0           1         280   
713              6           0       1970.0           2         576   
906              8           1       2006.0           3         865   

      WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  \
1389           0            0              0          0            0   
618            0          108              0          0          260   
227            0            0              0          0            0   
713            0          288              0          0            0   
906            0           60              0          0            0   

      PoolArea  MiscVal  MoSold  YrSold  
1389         0        0       3    2007  
618          0        0       7    2007  
227          0        0       9    2008  
713          0        0       5    2006  
906          0        0       6    2007

In [117]:
lmLasso_model.predict([[160,21.0,1869,6,6,1970,1970,127.0,321,0,162,483,483,504,0,987,0,0,1,1,2,1,5,0,1987.0,1,280,0,0,0,0,0,0,0,9,2008]])

array([109763.69655327])

In [118]:
new_test=pd.read_csv("/Users/rajkiran/Downloads/house-prices-advanced-regression-techniques/test.csv")
new_test

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave   NaN      Reg   
1     1462          20       RL         81.0    14267   Pave   NaN      IR1   
2     1463          60       RL         74.0    13830   Pave   NaN      IR1   
3     1464          60       RL         78.0     9978   Pave   NaN      IR1   
4     1465         120       RL         43.0     5005   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1454  2915         160       RM         21.0     1936   Pave   NaN      Reg   
1455  2916         160       RM         21.0     1894   Pave   NaN      Reg   
1456  2917          20       RL        160.0    20000   Pave   NaN      Reg   
1457  2918          85       RL         62.0    10441   Pave   NaN      Reg   
1458  2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0            Lvl    AllPub    Inside       Gtl        NAmes      Feedr   
1            Lvl    AllPub    Corner       Gtl        NAmes       Norm   
2            Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
3            Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
4            HLS    AllPub    Inside       Gtl      StoneBr       Norm   
...          ...       ...       ...       ...          ...        ...   
1454         Lvl    AllPub    Inside       Gtl      MeadowV       Norm   
1455         Lvl    AllPub    Inside       Gtl      MeadowV       Norm   
1456         Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
1457         Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
1458         Lvl    AllPub    Inside       Mod      Mitchel       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0          Norm     1Fam     1Story            5            6       1961   
1          Norm     1Fam     1Story            6            6       1958   
2          Norm     1Fam     2Story            5            5       1997   
3          Norm     1Fam     2Story            6            6       1998   
4          Norm   TwnhsE     1Story            8            5       1992   
...         ...      ...        ...          ...          ...        ...   
1454       Norm    Twnhs     2Story            4            7       1970   
1455       Norm   TwnhsE     2Story            4            5       1970   
1456       Norm     1Fam     1Story            5            7       1960   
1457       Norm     1Fam     SFoyer            5            5       1992   
1458       Norm     1Fam     2Story            7            5       1993   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0             1961     Gable  CompShg     VinylSd     VinylSd       None   
1             1958       Hip  CompShg     Wd Sdng     Wd Sdng    BrkFace   
2             1998     Gable  CompShg     VinylSd     VinylSd       None   
3             1998     Gable  CompShg     VinylSd     VinylSd    BrkFace   
4             1992     Gable  CompShg     HdBoard     HdBoard       None   
...            ...       ...      ...         ...         ...        ...   
1454          1970     Gable  CompShg     CemntBd     CmentBd       None   
1455          1970     Gable  CompShg     CemntBd     CmentBd       None   
1456          1996     Gable  CompShg     VinylSd     VinylSd       None   
1457          1992     Gable  CompShg     HdBoard     Wd Shng       None   
1458          1994     Gable  CompShg     HdBoard     HdBoard    BrkFace   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
0            0.0        TA        TA     CBlock       TA       TA   
1          108.0        TA        TA     CBlock       TA       TA   
2            0.0        TA        TA      PConc       Gd       TA   
3           20.0        TA        TA      PConc       TA       TA   

In [119]:
del new_test['Id']

In [120]:
new_test.shape

(1459, 79)

In [122]:
new_test_final = new_test.select_dtypes(['int64','float64'])

In [124]:
new_test_final['GarageYrBlt'].fillna(new_test_final['GarageYrBlt'].mean(),inplace=True)

In [125]:
new_test_final['LotFrontage'].fillna(new_test_final['LotFrontage'].mean(),inplace=True)

In [126]:
new_test_final['MasVnrArea'].fillna(new_test_final['MasVnrArea'].mean(),inplace=True)

In [127]:
def missing_zero_values_table(df):
        zero_val = (df == 0.00).astype(int).sum(axis=0)
        mis_val = df.isnull().sum()
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        mz_table = pd.concat([zero_val, mis_val, mis_val_percent], axis=1)
        mz_table = mz_table.rename(
        columns = {0 : 'Zero Values', 1 : 'Missing Values', 2 : '% of Total Values'})
        mz_table['Total Zero Missing Values'] = mz_table['Zero Values'] + mz_table['Missing Values']
        mz_table['% Total Zero Missing Values'] = 100 * mz_table['Total Zero Missing Values'] / len(df)
        mz_table['Data Type'] = df.dtypes
        mz_table = mz_table[
            mz_table.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns and " + str(df.shape[0]) + " Rows.\n"      
            "There are " + str(mz_table.shape[0]) +
              " columns that have missing values.")
#         mz_table.to_excel('D:/sampledata/missing_and_zero_values.xlsx', freeze_panes=(1,0), index = False)
        return mz_table

#https://stackoverflow.com/questions/51070985/find-out-the-percentage-of-missing-values-in-each-column-in-the-given-dataset

In [128]:
missing_zero_values_table(new_test_final)

Your selected dataframe has 36 columns and 1459 Rows.
There are 8 columns that have missing values.


Zero Values  Missing Values  % of Total Values  \
BsmtFullBath          849               2                0.1   
BsmtHalfBath         1364               2                0.1   
BsmtFinSF1            462               1                0.1   
BsmtFinSF2           1278               1                0.1   
BsmtUnfSF             123               1                0.1   
TotalBsmtSF            41               1                0.1   
GarageCars             76               1                0.1   
GarageArea             76               1                0.1   

              Total Zero Missing Values  % Total Zero Missing Values Data Type  
BsmtFullBath                        851                         58.3   float64  
BsmtHalfBath                       1366                         93.6   float64  
BsmtFinSF1                          463                         31.7   float64  
BsmtFinSF2                         1279                         87.7   float64  
BsmtUnfSF                           124                          8.5   float64  
TotalBsmtSF                          42                          2.9   float64  
GarageCars                           77                          5.3   float64  
GarageArea                           77                          5.3   float64

In [129]:
new_test_final['BsmtFullBath'].fillna(new_test_final['BsmtFullBath'].mean(),inplace=True)

In [130]:
new_test_final['BsmtHalfBath'].fillna(new_test_final['BsmtHalfBath'].mean(),inplace=True)

In [131]:
new_test_final['BsmtFinSF1'].fillna(new_test_final['BsmtFinSF1'].mean(),inplace=True)

In [132]:
new_test_final['BsmtFinSF2'].fillna(new_test_final['BsmtFinSF2'].mean(),inplace=True)

In [133]:
new_test_final['BsmtUnfSF'].fillna(new_test_final['BsmtUnfSF'].mean(),inplace=True)

In [134]:
new_test_final['TotalBsmtSF'].fillna(new_test_final['TotalBsmtSF'].mean(),inplace=True)

In [135]:
new_test_final['GarageCars'].fillna(new_test_final['GarageCars'].mean(),inplace=True)

In [136]:
new_test_final['GarageArea'].fillna(new_test_final['GarageArea'].mean(),inplace=True)

In [137]:
missing_zero_values_table(new_test_final)

Your selected dataframe has 36 columns and 1459 Rows.
There are 0 columns that have missing values.


Empty DataFrame
Columns: [Zero Values, Missing Values, % of Total Values, Total Zero Missing Values, % Total Zero Missing Values, Data Type]
Index: []

In [138]:
new_test_final.shape

(1459, 36)

In [171]:
lmLasso_model.predict(new_test_final)

array([109004.49200169, 148190.48937047, 167068.02513195, ...,
       164466.13502   ,  98502.54948261, 253917.71253045])

In [186]:
Predictions = lmLasso_model.predict(new_test_final)
type(Predictions)

numpy.ndarray

## In the above prediction, the 1st Predicted value : 109004's Lot area is 11662(this is test set).. now iam finding those houses whose lotarea  fall between 11660 -11665 in X_train... and Checking the price of the house in Train set ( better check in  TRAIN CSV FILE...)

In [220]:
X_train.loc[X_train['LotArea'].between(11650,11670) & X_train['OverallQual'].between(4,6) &\
            X_train['OverallCond'].between(5,7) & X_train['YearBuilt'].between(1900,1990) &\
           X_train['MSSubClass'].between(20,80)]

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
1014          20         60.0    11664            6            5       1948   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  \
1014          1950       206.0         336           0        746   

      TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
1014         1082      1082         0             0       1082             0   

      BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
1014             0         1         0             2             1   

      TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
1014             5           1       1948.0           1         240   

      WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  \
1014           0          130              0          0            0   

      PoolArea  MiscVal  MoSold  YrSold  
1014         0        0      11    2007

In [155]:
X_train.head()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
435           60         43.0    10667            7            6       1996   
1166          20         64.0    10475            8            5       2008   
935           30         52.0     5825            4            5       1926   
588           20         65.0    25095            5            8       1968   
80            60        100.0    13000            6            6       1968   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  \
435           1996         0.0         385         344         70   
1166          2008        72.0           0           0       1694   
935           1953       108.0           0           0        600   
588           2003         0.0        1324           0        113   
80            1968       576.0         448           0        448   

      TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
435           799       827       834             0       1661             1   
1166         1694      1694         0             0       1694             0   
935           600       747         0             0        747             0   
588          1437      1473         0             0       1473             2   
80            896      1182       960             0       2142             0   

      BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
435              0         2         1             3             1   
1166             0         2         0             3             1   
935              0         1         0             1             1   
588              0         1         0             1             1   
80               0         2         1             4             1   

      TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
435              6           1       1996.0           2         550   
1166             7           0       2008.0           3         776   
935              5           0       1953.0           2         528   
588              5           2       1968.0           1         452   
80               8           1       1968.0           1         509   

      WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  \
435          158           61              0          0            0   
1166         160           33              0          0            0   
935            0            0             32          0            0   
588            0           48              0          0           60   
80             0           72              0          0          252   

      PoolArea  MiscVal  MoSold  YrSold  
435          0        0       4    2009  
1166         0        0       2    2010  
935          0        0       6    2006  
588          0        0       6    2009  
80           0        0       6    2009

In [167]:
new_test_final

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0             20         80.0    11622            5            6       1961   
1             20         81.0    14267            6            6       1958   
2             60         74.0    13830            5            5       1997   
3             60         78.0     9978            6            6       1998   
4            120         43.0     5005            8            5       1992   
...          ...          ...      ...          ...          ...        ...   
1454         160         21.0     1936            4            7       1970   
1455         160         21.0     1894            4            5       1970   
1456          20        160.0    20000            5            7       1960   
1457          85         62.0    10441            5            5       1992   
1458          60         74.0     9627            7            5       1993   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  \
0             1961         0.0       468.0       144.0      270.0   
1             1958       108.0       923.0         0.0      406.0   
2             1998         0.0       791.0         0.0      137.0   
3             1998        20.0       602.0         0.0      324.0   
4             1992         0.0       263.0         0.0     1017.0   
...            ...         ...         ...         ...        ...   
1454          1970         0.0         0.0         0.0      546.0   
1455          1970         0.0       252.0         0.0      294.0   
1456          1996         0.0      1224.0         0.0        0.0   
1457          1992         0.0       337.0         0.0      575.0   
1458          1994        94.0       758.0         0.0      238.0   

      TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
0           882.0       896         0             0        896           0.0   
1          1329.0      1329         0             0       1329           0.0   
2           928.0       928       701             0       1629           0.0   
3           926.0       926       678             0       1604           0.0   
4          1280.0      1280         0             0       1280           0.0   
...           ...       ...       ...           ...        ...           ...   
1454        546.0       546       546             0       1092           0.0   
1455        546.0       546       546             0       1092           0.0   
1456       1224.0      1224         0             0       1224           1.0   
1457        912.0       970         0             0        970           0.0   
1458        996.0       996      1004             0       2000           0.0   

      BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
0              0.0         1         0             2             1   
1              0.0         1         1             3             1   
2              0.0         2         1             3             1   
3              0.0         2         1             3             1   
4              0.0         2         0             2             1   
...            ...       ...       ...           ...           ...   
1454           0.0         1         1             3             1   
1455           0.0         1         1             3             1   
1456           0.0         1         0             4             1   
1457           1.0         1         0             3             1   
1458           0.0         2         1             3             1   

      TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
0                5           0  1961.000000         1.0       730.0   
1                6           0  1958.000000         1.0       312.0   
2                6           1  1997.000000         2.0       482.0   
3                7           1  1998.000000         2.0       470.0   
4                5           0  1992.000000         2.0       506.0   


# Converting np Array to DF and Saving to .csv file

In [200]:
F_Pred = pd.DataFrame(Predictions,columns=['Predicted Sale Price']) #.to_csv("Preditions2.csv")

In [202]:
type(F_Pred)

pandas.core.frame.DataFrame

In [203]:
new_test_final_copy_2 = new_test_final

In [207]:
new_test_final_copy_df = pd.concat([new_test_final_copy_2,F_Pred],axis=1)

In [208]:
type(new_test_final_copy_df)

pandas.core.frame.DataFrame

In [212]:
new_test_final_copy_df.to_csv("Final Predictions With X's and Y.csv")  # Redded To Csv hahahaha...🤣🤣🤣🤣🤣🤣🤣

In [223]:
new_test_final_copy_df

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0             20         80.0    11622            5            6       1961   
1             20         81.0    14267            6            6       1958   
2             60         74.0    13830            5            5       1997   
3             60         78.0     9978            6            6       1998   
4            120         43.0     5005            8            5       1992   
...          ...          ...      ...          ...          ...        ...   
1454         160         21.0     1936            4            7       1970   
1455         160         21.0     1894            4            5       1970   
1456          20        160.0    20000            5            7       1960   
1457          85         62.0    10441            5            5       1992   
1458          60         74.0     9627            7            5       1993   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  \
0             1961         0.0       468.0       144.0      270.0   
1             1958       108.0       923.0         0.0      406.0   
2             1998         0.0       791.0         0.0      137.0   
3             1998        20.0       602.0         0.0      324.0   
4             1992         0.0       263.0         0.0     1017.0   
...            ...         ...         ...         ...        ...   
1454          1970         0.0         0.0         0.0      546.0   
1455          1970         0.0       252.0         0.0      294.0   
1456          1996         0.0      1224.0         0.0        0.0   
1457          1992         0.0       337.0         0.0      575.0   
1458          1994        94.0       758.0         0.0      238.0   

      TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
0           882.0       896         0             0        896           0.0   
1          1329.0      1329         0             0       1329           0.0   
2           928.0       928       701             0       1629           0.0   
3           926.0       926       678             0       1604           0.0   
4          1280.0      1280         0             0       1280           0.0   
...           ...       ...       ...           ...        ...           ...   
1454        546.0       546       546             0       1092           0.0   
1455        546.0       546       546             0       1092           0.0   
1456       1224.0      1224         0             0       1224           1.0   
1457        912.0       970         0             0        970           0.0   
1458        996.0       996      1004             0       2000           0.0   

      BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
0              0.0         1         0             2             1   
1              0.0         1         1             3             1   
2              0.0         2         1             3             1   
3              0.0         2         1             3             1   
4              0.0         2         0             2             1   
...            ...       ...       ...           ...           ...   
1454           0.0         1         1             3             1   
1455           0.0         1         1             3             1   
1456           0.0         1         0             4             1   
1457           1.0         1         0             3             1   
1458           0.0         2         1             3             1   

      TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  \
0                5           0  1961.000000         1.0       730.0   
1                6           0  1958.000000         1.0       312.0   
2                6           1  1997.000000         2.0       482.0   
3                7           1  1998.000000         2.0       470.0   
4                5           0  1992.000000         2.0       506.0   
